### PR TITLE
Documentation for the DI container

### PR DESCRIPTION
**Pull request by mnapoli**
I don't know if it's the right time to start documenting it. Feedback on the content is appreciated too.

### CODE REVIEW

**Comment by mattab** [comment]
our developer guide would only showcase those classes tagged with `@api`, maybe we are ready to tag Container and StaticContainer with `@api` ? 

in any case this guide would be very useful even for anyone contributing to Piwik core or plugins, and of course to community when writing custom plugins.

**Comment by mnapoli** [comment]
Your comment made me think about this and I actually wrote this doc for Piwik developers, not plugin developers.

I think we should hold off as long as we can exposing this to plugins, because as soon as it's exposed we have to keep BC. If it's only used in Piwik (and not 3rd party plugins) it leaves us some times to define best practices on naming container entries (in config files), on how to define entries, etc... We are just starting to use the container with plugins, we should gain more experience before freezing it with BC.

Also, `ContainerInterface` is a standard interface, not included in Piwik (and not related to PHP-DI). It doesn't make sense to tag it with `@api` since it's already an standard API (and it's not Piwik code anyway). `StaticContainer` is very bad practice, I don't want people using it, but at the same time Piwik's architecture doesn't leave us much of a choice… I'm conflicted.

So in conclusion, if we want the doc only for core developers we shouldn't merge it yet right?

**Comment by mattab** [comment]
> So in conclusion, if we want the doc only for core developers we shouldn't merge it yet right?

Correct, the developer.piwik.org is for Piwik users and plugin developers, so let's not merge it for now

**Comment by tsteur** [comment]
Now that we have a "Core Development" and a "Plugins / Develop" section can we write one guide for plugin developers under `/develop` Utils section that explains config and maybe DI to plugin developers. The guide should only contain information that is needed for plugin developers. And then an extended guide for core developers under `/core`? We could link from the short plugin guide to the extended guide if they want to know more.

**Comment by diosmosis** [comment]
I updated the PR. @mattab @tsteur please review and let me know if you have questions/comments.

File: docs/container.md [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+  * Implementations of the [Auth interface](http://developer.piwik.org/api-reference/Piwik/Auth).
+  * `Widgets` classes.
+  * Plugin `Settings` classes.
+  * `Type` classes.
+  * Plugin `Menu` classes.
+  * Plugin `Tasks` classes.
+
+  In 3.0, all classes will be constructed with dependency injection.
+
+### Configuring the container
```
Comment (tsteur) [inline docs/container.md]: I think all the part above this section should go into a different page under http://developer.piwik.org/piwik-in-depth as it is quite complicated and not really super relevant for plugin developers.

File: docs/container.md [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+We use dependency injection for the following reasons:
+
+* to move the responsibility of resolving dependencies outside of individual classes. Classes only list what they need, and the DI container gives it to them.
+* to allow plugins and users the ability to extend Piwik in any way they need by replacing and extending objects in the DI container.
+* to allow developers to more easily test classes both in isolation and in the Piwik environment, by making it easy to inject mocks.
+* to allow developers to more easily maintain code modularity by avoiding hard-coded dependencies.
+
+### What classes go in the container
+
+In Piwik development, there are four types of classes that exist:
```
Comment (tsteur) [inline docs/container.md]: Is this part kinda repeating the first part? Maybe also rather something for  a page under http://developer.piwik.org/piwik-in-depth ?

File: docs/container.md [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+  * `Widgets` classes.
+  * Plugin `Settings` classes.
+  * `Type` classes.
+  * Plugin `Menu` classes.
+  * Plugin `Tasks` classes.
+
+  In 3.0, all classes will be constructed with dependency injection.
+
+### Configuring the container
+
```
Comment (tsteur) [inline docs/container.md]: I think for plugin developers relevant is 

* How to configure the DI container and how to replace dependencies etc (covered by this section)
* How to get an instance from DI container in the plugin itself (eg via constructor). Maybe we could also explain why one something has to use `StaticContainer` etc. Covered partially by the other sections but too detailed and too complicated I'd say. Maybe we can use that example `API` above to explain it.

Plugins should be pretty small usually and DI might be not too important for them. Eg they possibly don't need to create classes that can be changed by other plugins etc. This would be rather a use case for the core. I think for plugins more important is how can I change dependencies defined in core and how can I get the correct dependency eg for `Config`, `DB` in core

File: docs/container.md [hunk 4]
```diff
@@ -0,0 +1,313 @@
+---
+category: Develop
+previous: piwiks-ini-configuration
+next: plugin-settings
```
Comment (tsteur) [inline docs/container.md]: I'm not sure where this would currently appear as `piwiks-ini-configuration` is under "Piwik In Depth" and `plugin-settings` under "Develop > Utils". 

Maybe we could have
* one guide `Dependency Injection` under "Develop > Utils" explaining how to get dependencies
* one guide `Configuration` under "Develop > Plugin Basics" describing the config section from the guide below and all the possibilities one has. This could be as well part of the `Dependency Injection` guide. I'm not sure if it's beneficial to have 2 separate guides for this or whether it is confusing. We can still link between them so one can find it in 2 places which might be better. Not sure!
* and the more detailed documentation part in "Piwik In Depth > Utils" next or in the configuration page?

File: docs/container.md (hunk 3 above)
Comment (diosmosis) [inline docs/container.md]: This guide is mostly for core + PRO developers. 3rd party plugin developers do not actually have to use DI, and if they never write tests they're obviously never going to use DI. I can move the whole guide to piwik-in-depth or to some section for core/pro devs and revert to napoli's original guide for plugin devs.

> Eg they possibly don't need to create classes that can be changed by other plugins etc.

I know this is about plugin development, but I want to make it clear in case it's not: for core development (and probably pro development, though I don't know how much affect we have there), all classes that contain important application logic should be stored in the container. Regardless of whether there is a reason for them to be changed by plugins or not.

It is not each class' responsibility to obtain their dependencies (ie, by creating instances manually, accessing singletons or using a service locator), it is the container's responsibility to wire them. Any class that doesn't have important application logic could be replaced or stored in a 3rd party library or component.

File: docs/container.md (hunk 4 above)
Comment (diosmosis) [inline docs/container.md]: This sounds good to me. Will look into it when I have the time.

> I'm not sure if it's beneficial to have 2 separate guides for this or whether it is confusing. We can still link between them so one can find it in 2 places which might be better. Not sure!

I think it's ok, some redundancy in the writing (specific to each use case) could provide a greater chance the reader will understand.

**Comment by tsteur** [comment]
documented in https://github.com/matomo-org/developer-documentation/pull/356 and targeted for plugin developers
